charts/kubearchive/templates/api_server/api_server.yaml

```
@@ -49,6 +49,10 @@ spec:
             - name: POSTGRES_PORT
               value: "{{ required "A Postgres port must be specified" .Values.database.service.port }}"
             {{- include "kubearchive.v1.otel.env" . | nindent 12 }}
+            - name: CACHE_EXPIRATION_AUTHORIZED
+              value: "10m"
+            - name: CACHE_EXPIRATION_UNAUTHORIZED
+              value: "1m"
```

Although this is perfectly valid, I would go with not setting the env vars by default, probably something like this:

```suggestion
           {{- if .Values.apiServer.expirationAuth }}
            - name: CACHE_EXPIRATION_AUTHORIZED
              value: "10m"
            {{- end }}
            {{- if .Values.apiServer.cache.expirationUnauth }}
            - name: CACHE_EXPIRATION_UNAUTHORIZED
              value: "1m"
            {{-end}}
```

And add a couple of comments in the values.yaml with an example of how to override the default values stating which is the default when unset.

---

I don't think that moving the default value to the Values file provides value. I think the difference between production and development comes into play here. As this is something configurable by the user in production, it makes sense that we configure it in the same way in development.

---

Let me rephrase this: 

I think it makes more sense having the default cache expiration times set in code so when no environment variables are provided, these values are in place.

Now, for how to better communicate the developers/users that this values can be overwritten:
* For the users, as you already did, having instructions for how to introduce this env vars in the Deployment
* For the developers, as we have the helm chart, I think that if we centralize this kind of stuff in the `Values.yaml` instead of having to open the different templates provides value.

---

I'm OK with moving the default values to the code.

I don't agree with having the option to set the values in development using the Values.yaml. I do think it contributes to distance the development experience from the production experience, which is something I think we need to keep to a minimum. However I does not bother me that the option is available to whoever wants to use it.

---

the reason I proposed to use the helm templated values is because we have helm in place

if we want to narrow the gap between experience of the user and the developer by not using helm templated values, does it makes sense to get rid of helm at some point?

---

I think that ideally we should be deploying in development as we deploy in production, but the world is not ideal. So I understand that helm might be useful to auto-deploy some large stuff that is not deployed on production, in our case the database deployment or a telemetry deployment. For configuration that is offered to the user I find more useful to use the production process so I can feel something similar to what they are feeling and it is what I do.

So while I think it makes sense to get rid of Helm, I don't think we want to get rid of it ever, as it is helpful to us. And given that, and given that I personally can deploy without using Helm, its fine for me to add more stuff into Helm.

---

your statement makes a lot of sense, and it actually could be a design decision that we are explicitly making for "How a user should deploy KubeArchive and why do we include Helm charts just for development"

---

> I think it makes more sense having the default cache expiration times set in code so when no environment variables are provided, these values are in place.

Why?  Will the container ever be run without the deployment? It seems reasonable to me, and common practice, to have environment variables with defaults.  

I don't think *everything* has to go into values.yaml.

---

> Why? Will the container ever be run without the deployment? It seems reasonable to me, and common practice, to have environment variables with defaults.

I am more comfortable with the idea of not needing environment variables and then, if you look into customization of the api, override the default values with environment variables rather than modifying them.

If I thought that the cache expiration time is going to be something that each user is going to change, I would prefer it to be more explicit, and to fail if no env variable is provided, but I think that's not the case and it makes sense to don't care about those parameters in general.

With this, the YAML provided to be deployed is shorter and cleaner.

> I don't think everything has to go into values.yaml.
I don't think that everything has to be in the `values.yaml` either, but the things that could be changed yes, and for development purposes I think that I may want to be able to change the cache expiration time, and everything I want to tweak regarding the kubearchive configuration is in one place, I like it better than scattered in different files

---

> but the things that could be changed yes

This is where I think we disagree. Just the fact that something *could* be changed doesn't mean it has to go into the `values.yaml` file.  There are a million things that *could* be changed.

But if something is not likely to change, I don't see any value in putting it in `values.yaml`.  And I don't think these are likely to change.

IMO, this will also complicate debugging things in production.  A user would then have to go find the variable names and add them to the deployment, hopefully getting the names correct and doing it right. If they're there, he just has to change the value. You don't really need documentation for that.

---

> This is where I think we disagree. Just the fact that something could be changed doesn't mean it has to go into the values.yaml file. There are a million things that could be changed.

You are right, and as you say, it makes more sense to say that we should add to the `values.yaml` the parameters that are likely to change by the developers (as we are using helm charts for development)

But my point still stands because as I mentioned earlier, for development purposes I believe I may want to change those parameters.

> IMO, this will also complicate debugging things in production. A user would then have to go find the variable names and add them to the deployment, hopefully getting the names correct and doing it right. If they're there, he just has to change the value. You don't really need documentation for that.

About this, they are different perspectives, I prefer a good documentation an smaller kubernetes resources for things that are optional when we talk about providing things to the user, you prefer having every option explicitly set so the user won't need to look for the docs to customize optional stuff.

If you are worried about having to look for the environment variable name, we could also add a comment in the deployment with the line that should be uncommented and edited to add the env var.

```yaml
           {{- if .Values.apiServer.expirationAuth }}
            - name: CACHE_EXPIRATION_AUTHORIZED
              value: "10m"
            {{- else }}
            # Uncomment the following lines updating the value to override the default value of 10 minutes
            # Check the docs for more information
            # - name: CACHE_EXPIRATION_AUTHORIZED
            #   value: "10m"
            {{- end }}
            {{- if .Values.apiServer.cache.expirationUnauth }}
            - name: CACHE_EXPIRATION_UNAUTHORIZED
              value: "1m"
            {{- else }}
            # Uncomment the following lines updating the value to override the default value of 1 minute. 
            # Check the docs for more information
            # - name: CACHE_EXPIRATION_UNAUTHORIZED
            #   value: "1m"
            {{- end }}
```